cli/azd/pkg/tools/git/git.go

```diff
@@ -245,3 +248,58 @@ func (cli *gitCli) IsUntrackedFile(ctx context.Context, repositoryPath string, f
 
 	return false, nil
 }
+
+// SetGitHubAuthForRepo creates git config for the repositoryPath like
+//
+// [credential "https://github.com"]  (when credential is equal to "https://github.com")
+// helper =
+// helper = !ghPath auth git-credential
+//
+// This way, git commands run from repositoryPath will use gh as auth credential.
+// Note: Removes any previous configuration for the credential.
+// Note: `helper = ` is intentional to break the chain of previously configured global helpers.
+// See: https://github.com/cli/cli/issues/3796 for more about this strategy.
+func (cli *gitCli) SetGitHubAuthForRepo(ctx context.Context, repositoryPath, credential, ghPath string) error {
+
+	if err := setAuthCredentialHelper(
+		ctx, cli.commandRunner, repositoryPath, credential, "", "replace-all"); err != nil {
+		return err
+	}
+	// path needs to be quoted on windows
+	if runtime.GOOS == "windows" {
+		ghPath = fmt.Sprintf("'%s'", ghPath)
+	}
+	ghCredentialValue := fmt.Sprintf("!%s auth git-credential", ghPath)
+	if err := setAuthCredentialHelper(
+		ctx, cli.commandRunner, repositoryPath, credential, ghCredentialValue, "add"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func setAuthCredentialHelper(
+	ctx context.Context, runner exec.CommandRunner, repositoryPath, credential, value, flag string) error {
+	runArgs := newRunArgs(
+		"-C", repositoryPath,
+		"config", fmt.Sprintf("--%s", flag),
+		fmt.Sprintf("credential.%s.helper", credential),
+		value)
+
+	if _, err := runner.Run(ctx, runArgs); err != nil {
+		return fmt.Errorf("failed to set credential helper: %s='%s': %w", credential, value, err)
+	}
+	return nil
+}
+
+func newRunArgs(args ...string) exec.RunArgs {
+
+	runArgs := exec.NewRunArgs("git", args...)
+	if os.Getenv("CODESPACES") == "true" {
+		// azd running git in codespaces should not use the Codespaces token.
+		// As azd needs bigger access across repos. And the token in codespaces is mono-repo by default
+		runArgs = runArgs.WithEnv([]string{"GITHUB_TOKEN=", "GH_TOKEN="})
+	}
```

Comment:
@vhvb1989 The original implementation (which was regressed due to treating the `cli` object as a temp-scoped object) worked in a way that wasn't tied to `CODESPACES`. The implementation was perhaps a "smarter" detection, but I could also see us needing a more heavy hammer for saying let's just make something that always works reasonably for `CODESPACES`. I can see going either way, but just want to hear thoughts on the tradeoffs.

Reply:
> which was regressed due to treating the cli object as a temp-scoped object

I think the regression happened due to bringing a spinner on top of the `setting github secret` operation.  I never considered `gh secret set foo --body bar` to be `interactive`, as the gh cli would either do it of fail. I missed that we were re-launching `gh auth login` on set secret failed do to `not enough access`.

So, I could have just stopped the spinner for setting a secret and resume it after it... to allow the interaction and have the same functionality than before. But we were only covering the `set secret operation`.  The other common issue was the common inability to `push changes` to the just created repo...  because `git` is using the authentication provided by codespaces which is limited to read/write only for the current repo.

So my approach in the PR is to completelly ignore codespaces default auth when using git or ghcli from azd (basically for azd pipelien config).  This makes the `gh cli login` to be the first step from running `azd pipeline config` (compared to running login in the middle of the process, after selecting or creating a scm repo).

Reply:
> What is the expectation when the user is not in codespaces and they have an incorrect/invalid token value in GITHUB_TOKEN ??

@wbreza the expectation is to see azd fail if the token in GITUB_TOKEN has not enough access to perform the operation.  The error would be coming out of gh cli as `not authorized`.